File: Day-15/main.cpp
#include <list> 
#include <iostream>
#include <unordered_map>

using namespace std;

int simulate(list<int> l, int totalTurns) {
	unordered_map<int,int> dict;

	// Insert pair (i,0) for every i in mem except the last one
	int i=1;
	for(list<int>::iterator it=l.begin(); it != --l.end(); ++it){
	    dict.insert(std::make_pair(*it, i));
	    i++;
	}

	int lastSpoken = *(--l.end()); // Last entry of l
	for(int t=l.size()+1; t<=totalTurns; t++) {
		int temp = lastSpoken;
		if(dict.find(lastSpoken) == dict.end()) {
			dict[lastSpoken] = t-1;
			lastSpoken = 0;			
		} else {
			lastSpoken = (t-1) - dict[lastSpoken];
		}

		dict[temp] = t-1;
	}
	return lastSpoken;
}

int main() {
	list<int> l = {13,16,0,12,15,1};
	cout << "Part 1: " << endl << simulate(l, 2020) << endl;
	cout << "Part 2: " << endl << simulate(l, 30000000) << endl; // Can take up to 30sec
	return 0;
}

File: Day-15/main.py
def readData(filename):
	with open(filename, "r") as file:
		data = file.read().split(",")
		return list(map(lambda x: int(x), data))

def simulate(numbers, totalTurns):
	dictionary = dict()
	numbersLength = len(numbers)
	for i in range(len(numbers)-1):
		dictionary[numbers[i]] = i+1


	lastSpoken = numbers[numbersLength-1]
	for t in range(numbersLength+1, totalTurns+1):
		temp = lastSpoken
		if lastSpoken in dictionary:
			lastSpoken = (t-1) - dictionary[lastSpoken]
		else:
			dictionary[lastSpoken] = t-1
			lastSpoken = 0
		dictionary[temp] = t-1
	return lastSpoken


if __name__ == "__main__":
	numbers = readData("input.txt")
	print("Starting Numbers:", numbers)

	print(simulate(numbers, 2020))
	print(simulate(numbers, 30000000)) # Can take up to 30sec
